pkg/check/cache.go

```diff
@@ -0,0 +1,282 @@
+}
+
+func (c *vSphereCache) GetDatastoreByURL(ctx context.Context, dcName, dsURL string) (mo.Datastore, error) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	cdss, err := c.getDatastoresLocked(ctx, dcName)
+	if err != nil {
+		return mo.Datastore{}, err
+	}
+
+	for _, cds := range cdss {
+		if cds.dsMo.Info.GetDatastoreInfo().Url == dsURL {
+			klog.V(4).Infof("Found datastore MoRef %v for datastoreURL: %q in datacenter: %q",
+				cds.dsMo.Reference(), dsURL, dcName)
+			return *cds.dsMo, nil
+		}
+	}
+	err = fmt.Errorf("couldn't find Datastore given URL %q in datacenter %s", dsURL, dcName)
+	klog.Error(err)
+	return mo.Datastore{}, err
+}
+
+func (c *vSphereCache) GetDatastoreMo(ctx context.Context, dcName, dsName string) (mo.Datastore, error) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	cdss, err := c.getDatastoresLocked(ctx, dcName)
+	if err != nil {
+		return mo.Datastore{}, err
+	}
+
+	for _, cds := range cdss {
+		// some callers use name, others use path
+		if cds.ds.Name() == dsName || cds.dsPath == dsName {
+			return *cds.dsMo, nil
+		}
+	}
+	err = fmt.Errorf("couldn't find Datastore named %q in datacenter %s", dsName, dcName)
+	klog.Error(err)
+	return mo.Datastore{}, err
+}
+
+func (c *vSphereCache) GetDatastoreMoByReference(ctx context.Context, dcName string, ref types.ManagedObjectReference) (mo.Datastore, error) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	cdss, err := c.getDatastoresLocked(ctx, dcName)
+	if err != nil {
+		return mo.Datastore{}, err
+	}
+
+	for _, cds := range cdss {
+		// some callers use name, others use path
+		if cds.ds.Reference() == ref {
+			return *cds.dsMo, nil
+		}
+	}
+	err = fmt.Errorf("couldn't find Datastore with reference %+v in datacenter %s", ref, dcName)
+	klog.Error(err)
+	return mo.Datastore{}, err
+}
+
+func (c *vSphereCache) GetStoragePods(ctx context.Context) ([]mo.StoragePod, error) {
+	c.mutex.Lock()
+	defer c.mutex.Unlock()
+
+	if c.storagePods != nil {
+		return c.storagePods, nil
+	}
+
+	// list datastore cluster
+	m := view.NewManager(c.vmClient)
+	kind := []string{"StoragePod"}
+	tctx, cancel := context.WithTimeout(ctx, *Timeout)
+	defer cancel()
+	v, err := m.CreateContainerView(tctx, c.vmClient.ServiceContent.RootFolder, kind, true)
+	if err != nil {
+		klog.Errorf("error listing datastore cluster: %+v", err)
+		// Don't alert on missing permissions
+		return nil, nil
+	}
+	defer func() {
+		v.Destroy(tctx)
+	}()
+
+	var content []mo.StoragePod
+	tctx, cancel = context.WithTimeout(ctx, *Timeout)
+	defer cancel()
+	err = v.Retrieve(tctx, kind, []string{SummaryProperty, "childEntity"}, &content)
+	if err != nil {
+		klog.Errorf("error retrieving datastore cluster properties: %+v", err)
+		// it is possible that we do not actually have permission to fetch datastore clusters
+		// in which case rather than throwing an error - we will silently return nil, so as
+		// we don't trigger unnecessary alerts.
+		return nil, nil
+	}
+	c.storagePods = content
+	return content, nil
+}
```

Comment:
Do we need some tests for this?

Reply:
I added some tests that shut down the simulator